File: nodejs/expressmvc/routes/receita-simples-express.js
const express = require('express');
const path = require('path');
const uuidv4 = require('uuid/v4');

const {
	PadesSignatureStarter,
	StandardSignaturePolicies,
	SignatureFinisher,
} = require('pki-express');

const { Util } = require('../util');
const { Prescricao } = require('../prescricao');
const { StorageMock } = require('../storage-mock');
const { PadesVisualElementsExpress } = require('../pades-visual-elements-express');

const router = express.Router();
const APP_ROOT = process.cwd();

/**
 * GET /receita-simples-express/
 * 
 * This file initiates the generation of Receita Simples file.
 * 
 */
router.get('/', async (_, res) => {
	const ufs = ['AC', 'AL', 'AM', 'AP', 'BA', 'CE', 'DF', 'ES', 'GO', 'MA', 'MG', 'MS', 'MT', 
	'PA', 'PB', 'PE', 'PI', 'PR', 'RJ', 'RN', 'RO', 'RR', 'RS', 'SC', 'SE', 'SP', 'TO'];

	res.render('receita-simples-express', { ufs });
});

/**
 * POST /receita-simples-express/generate
 *
 * This file generates the Receita Simples file and it performs a 
 * PAdES signature in three steps using PKI Express and Web PKI.
 * 
 */
router.post('/generate', async (req, res, next) => {
	// Recover variables from the POST arguments to be use don this step.
	const name = req.body.name;
	const crm = req.body.crm;
	const crmUf = req.body.crmUf;
	
	// Generate Receita Simples
	const fileId = Prescricao.generateReceitaSimples(name, crm, crmUf);

	res.render('receita-simples-express/generate', { fileId });
});

/**
 * POST /receita-simples-express/start
 *
 * This action receives the form submission from the signature page. 
 * It will perform a PAdES signature in three steps using PKI Express
 * and Web PKI.
 * 
 */
router.post('/start', (req, res, next) => {
	const { fileId } = req.query;

	// Recover variables from the POST arguments to be use don this step.
	const certThumb = req.body.certThumbField;
	const certContent = req.body.certContentField;

	// Get an instantiate of the PadesSignatureStarter class, responsible for
	// receiving the signature elements and start the signature process.
	const signatureStarter = new PadesSignatureStarter();

	// Set PKI default options (see util.js).
	Util.setPkiDefaults(signatureStarter);

	// Set PDF to be signed.
	signatureStarter.setPdfToSignFromPathSync(StorageMock.getDataPath(fileId));

	// REQUIRED!
	// Use a custom signature field name. This field MUST have the 
	// "Emitente" keyword as the last keyword.
	signatureStarter.customSignatureFieldName = "Signature1 Emitente";

	// Set signature policy.
	signatureStarter.signaturePolicy = StandardSignaturePolicies.PADES_BASIC;
	
	// Set Base64-encoded certificate's content to signature starter.
	signatureStarter.setCertificateFromBase64Sync(certContent);

	// Set a file reference for the stamp file. Note that this file can be
	// referenced later by "fref://{alias}" at the "url" field on the
	// visual representation (see public/vr.json or
	// getVisualRepresentation() method).
	signatureStarter.addFileReferenceSync('stamp', StorageMock.getPdfStampPath());

	// Set the visual representation. We provided a dictionary that
	// represents the visual representation JSON model.
	signatureStarter
		.setVisualRepresentationSync(PadesVisualElementsExpress.getVisualRepresentation());

	// Start the signature process.
	signatureStarter.start()
		.then((response) => {
			// Receive as response the following fields:
			// - toSignHash: The hash to be signed.
			// - digestAlgorithm: The digest algorithm that will inform the Web PKI
			//                    component to compute this signature.
			// - transferFile: A temporary file to be passed to "complete" step.
			const { toSignHash } = response;
			const { digestAlgorithm } = response;
			const { transferFile } = response;

			// Render the field from start() method as hidden field to be used on
			// the javascript or on the "complete" step.
			res.render('receita-simples-express/start', {
				toSignHash,
				digestAlgorithm,
				transferFile,
				certThumb,
				fileId,
			});
		}).catch((err) => next(err));
});

/**
 * POST receita-simples-express/complete
 *
 * This route completes the signature, it will be called programatically after
 * the Web PKI component perform the signature and submit the form (see method
 * sign() on public/javascripts/signature-complete-form.js).
 */
router.post('/complete', (req, res, next) => {
	const { fileId } = req.query;

	// Recover variables from the POST arguments to be used on this step.
	const transferFile = req.body.transferFileField;
	const signature = req.body.signatureField;

	// Get an instance of the PadesSignatureFinisher class, responsible for
	// completing the signature process.
	const signatureFinisher = new SignatureFinisher();

	// Set PKI default options (see util.js).
	Util.setPkiDefaults(signatureFinisher);

	// Set PDF to be signed. It's the same file we used on "start" step.
	signatureFinisher.setFileToSignFromPathSync(StorageMock.getDataPath(fileId));

	// Set transfer file.
	signatureFinisher.setTransferFileFromPathSync(transferFile);

	// Set signature.
	signatureFinisher.signature = signature;

	// Generate path for output file and add the signature finisher.
	StorageMock.createAppDataSync(); // Make sure the "app-data" folder exists.
	const outputFile = `${uuidv4()}.pdf`;
	signatureFinisher.outputFile = path.join(APP_ROOT, 'app-data', outputFile);

	// Complete the signature process.
	const getCert = true;
	signatureFinisher
		.complete(getCert)
		.then((result) => {
			// After complete the signature, render the result page, passing the
			// outputFile containing the signed file.
			const certificate = result;
			res.render('receita-simples-express/complete', {
				signedPdf: outputFile,
				signerCert: certificate,
			});
		})
		.catch((err) => next(err));
});

module.exports = router;

File: python/flask/sample/views/receita_simples_express.py
import uuid
import os

from flask import request
from flask import Blueprint
from flask import current_app
from flask import render_template
from flask import request
from flask import current_app
from pkiexpress import standard_signature_policies
from pkiexpress import PadesSignatureStarter
from pkiexpress import SignatureFinisher

from sample.pades_visual_elements_express import PadesVisualElementsExpress

from sample.prescricao import generate_receita_simples

from sample.storage_mock import create_app_data
from sample.utils import set_pki_defaults
from sample.storage_mock import get_pdf_stamp_path


blueprint = Blueprint('receita_simples_express',__name__, 
                      url_prefix='/receita-simples-express')

@blueprint.route('/', methods=['GET'])
def index():
    ufs=["AC", "AL", "AM", "AP", "BA", "CE", "DF", "ES", "GO", "MA", "MG", "MS", "MT", "PA",
			 "PB", "PE", "PI", "PR", "RJ", "RN", "RO", "RR", "RS", "SC", "SE", "SP", "TO"]
    
    # Render the signature page (templates/receita_simples_express/index.html)
    return render_template('receita_simples_express/index.html', ufs=ufs)


@blueprint.route('/sign', methods=['GET', 'POST'])
def post_sign():
    try:
    # Recover name, crm and crmUf from the POST argument.
        name = request.form['name']
        crm = request.form['crm']
        crmUf = request.form['crmUf']
    except Exception as e:
        return render_template('error.html', msg=e)
  
    # Generate a new PDF document, sent as a string stream, with the given information
    file_stream = generate_receita_simples(name, crm, crmUf)
    file_extension = "pdf"

    # Generate a unique filename.
    file_id = '%s.%s' % (str(uuid.uuid4()), file_extension)

    # Move the file to the "app_data" with the unique filename. Make sure
    # the "app_data" folder exists (static/util.py).
    create_app_data()
    with open(os.path.join(current_app.config['APPDATA_FOLDER'], file_id), 'wb') as f:
        f.write(file_stream)

    return render_template('receita_simples_express/sign.html', file_id=file_id)


@blueprint.route('/start/<file_id>', methods=['POST'])
def start(file_id):
    """

    This method starts the signature. In this sample, it will be called
    programatically after the user press the "Sign File" button (see method
    readCertificate() on static/js/signature-start-form.js).

    """
    try:
        # Recover variables from the POST arguments to be used on this step.
        cert_thumb = request.form['certThumbField']
        cert_content = request.form['certContentField']

        # Get an instance of the PadesSignatureStarter class, responsible for
        # receiving the signature elements and start the signature process.
        signature_starter = PadesSignatureStarter()

        # Set PKI default options (see utils.py).
        set_pki_defaults(signature_starter)

        # Set signature policy.
        signature_starter.signature_policy = \
            standard_signature_policies.PADES_BASIC

        # Set PDF to be signed
        signature_starter.set_pdf_to_sign_from_path(
            os.path.join(current_app.config['APPDATA_FOLDER'], file_id))

        # Set Base64-encoded certificate's content to signature starter.
        signature_starter.set_certificate_from_base64(cert_content)

        # Set a file reference for the stamp file. Note that this file can be
        # referenced later by "fref://{alias}" at the "url" field on the visual
        # representation (see static/vr.json or get_visual_representation()
        # method).
        signature_starter.add_file_reference('stamp', get_pdf_stamp_path())

        # Set the visual representation. We provided a dictionary that
        # represents the visual representation JSON model.
        signature_starter.set_visual_representation(
            PadesVisualElementsExpress.get_visual_representation())

        # Start the signature process. Receive as response the following fields:
        # - to_sign_hash:     The hash to be signed.
        # - digest_algorithm: The digest algorithm that will inform the Web PKI
        #                     component to compute the signature.
        # - transfer_file:    A temporary file to be passed to "complete" step.
        response = signature_starter.start()

        # Render the field from start() method as hidden field to be used on the
        # javascript or on the "complete" step.
        return render_template('receita_simples_express/start.html',
                               cert_thumb=cert_thumb,
                               transfer_file_id=response['transferFile'],
                               to_sign_hash=response['toSignHash'],
                               digest_algorithm=response['digestAlgorithm'])

    except Exception as e:
        return render_template('error.html', msg=e)

@blueprint.route('/complete/<file_id>', methods=['POST'])
def complete(file_id):
    """

    This function completes the signature, it will be called programatically
    after the Web PKI component perform the signature and submit the form (see
    method sign() on static/js/signature-complete-form.js).

    """
    try:

        # Recover variables form the POST arguments to be used on this step.
        transfer_file = request.form['transferFileIdField']
        signature = request.form['signatureField']

        # Get an instance of the SignatureFinisher class, responsible for
        # completing the signature process.
        signature_finisher = SignatureFinisher()

        # Set PKI default options (see utils.py).
        set_pki_defaults(signature_finisher)

        # Set PDF to be signed. It's the same file we used on "start" method.
        signature_finisher.set_file_to_sign_from_path(
            os.path.join(current_app.config['APPDATA_FOLDER'], file_id))

        # Set the transfer file.
        signature_finisher.set_transfer_file_from_path(transfer_file)

        # Set the signature file.
        signature_finisher.signature = signature

        # Generate path for output file and add to the signature finisher.
        create_app_data()  # Guarantees that "app data" folder exists.
        filename = '%s.pdf' % (str(uuid.uuid4()))
        signature_finisher.output_file = \
            os.path.join(current_app.config['APPDATA_FOLDER'], filename)

        # Complete the signature process.
        signer_cert = signature_finisher.complete(get_cert=True)

        return render_template('receita_simples_express/signature-info.html',
                               signer_cert=signer_cert,
                               signed_pdf=filename)

    except Exception as e:
        return render_template('error.html', msg=e)